src/topology/maps.lean

```diff
@@ -13,130 +13,200 @@ local attribute [instance] classical.prop_decidable
 
 variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
 
+section dense_range
+variables [topological_space α] [topological_space β] [topological_space γ]
+          (f : α → β) (g : β → γ)
+
+def dense_range := ∀ x, x ∈ closure (range f)
+
+lemma dense_range_iff_closure_eq : dense_range f ↔ closure (range f) = univ :=
+eq_univ_iff_forall.symm
+
+variables {f}
+
+lemma dense_range_comp (hf : dense_range f) (hg : dense_range g) (cg : continuous g) :
+  dense_range (g ∘ f) :=
+begin
+  have : g '' (closure $ range f) ⊆ closure (g '' range f),
+    from image_closure_subset_closure_image cg,
+  have : closure (g '' closure (range f)) ⊆ closure (g '' range f),
+    by simpa [closure_closure] using (closure_mono this),
+  intro c,
+  rw range_comp,
+  apply this,
+  rw [(dense_range_iff_closure_eq f).1 hf, image_univ],
+  exact hg c
+end
+
+lemma dense_range.inhabited (df : dense_range f) (b : β) : inhabited α :=
```

Comment:
Why `inhabited` instead of `nonempty`? since there is no canonical element, I think `nonempty` is more natural -- and if you really need `inhabited` for some applications, maybe it means that the statements you rely on should also use `nonempty` instead of `inhabited`, or otherwise you can easily derive `inhabited` from `nonempty` in the application.

Reply:
`inhabited` is needed because this is what `lim` wants.